model/waterfall.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+const (
+	DefaultWaterfallVersionLimit = 5
+	MaxWaterfallVersionLimit     = 300
+)
+
+type WaterfallTask struct {
+	Id          string `bson:"_id" json:"_id"`
+	DisplayName string `bson:"display_name" json:"display_name,omitempty"`
+	Status      string `bson:"status" json:"status,omitempty"`
+}
+
+type WaterfallBuild struct {
+	Id          string          `bson:"_id" json:"_id"`
+	Activated   bool            `bson:"activated" json:"activated,omitempty"`
+	DisplayName string          `bson:"display_name" json:"display_name,omitempty"`
+	Version     string          `bson:"version" json:"version,omitempty"`
+	Tasks       []WaterfallTask `bson:"tasks" json:"tasks,omitempty"`
+}
+
+type WaterfallBuildVariant struct {
+	Id          string           `bson:"_id" json:"_id"`
+	DisplayName string           `bson:"display_name" json:"display_name,omitempty"`
+	Builds      []WaterfallBuild `bson:"builds" json:"builds,omitempty"`
+}
+
+type WaterfallOptions struct {
+	Limit      int
+	Requesters []string
+}
+
+// GetWaterfallVersions returns `opts.limit` versions for a given project.
+// TODO: It will eventually return `opts.limit` activated versions that also satisfy any given filters.
+func GetWaterfallVersions(ctx context.Context, projectId string, opts WaterfallOptions) ([]Version, error) {
+	invalidRequesters, _ := utility.StringSliceSymmetricDifference(opts.Requesters, evergreen.SystemVersionRequesterTypes)
+	if len(invalidRequesters) > 0 {
+		return nil, errors.Errorf("invalid requesters '%s'", invalidRequesters)
+	}
+	match := bson.M{
+		VersionIdentifierKey: projectId,
+		VersionRequesterKey: bson.M{
+			"$in": opts.Requesters,
+		},
+	}
+
+	// TODO DEVPROD-10177: Add revision order logic to handle pagination.
+
+	pipeline := []bson.M{{"$match": match}}
+	pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+	limit := defaultVersionLimit
+	if opts.Limit != 0 {
+		limit = opts.Limit
+	}
+
+	pipeline = append(pipeline, bson.M{"$limit": limit})
+
+	res := []Version{}
+	env := evergreen.GetEnvironment()
+	cursor, err := env.DB().Collection(VersionCollection).Aggregate(ctx, pipeline)
+	if err != nil {
+		return nil, errors.Wrap(err, "aggregating versions")
+	}
+	err = cursor.All(ctx, &res)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// GetWaterfallBuildVariants returns all build variants associated with the specified versions. Each build variant contains an array of builds sorted by revision and their tasks.
+func GetWaterfallBuildVariants(ctx context.Context, versions []Version) ([]WaterfallBuildVariant, error) {
+	versionIds := []string{}
+	for _, version := range versions {
+		versionIds = append(versionIds, version.Id)
+	}
+
+	pipeline := []bson.M{{"$match": bson.M{VersionIdKey: bson.M{"$in": versionIds}}}}
```

Comment:
nit: like above, might help to check that `len(versionIds) > 0` before running the query.